compiler/src/org.graalvm.compiler.loop.phases/src/org/graalvm/compiler/loop/phases/LoopPredicationPhase.java

```diff
@@ -0,0 +1,282 @@
+                    assert !workList.contains(frameState);
+                    workList.push(frameState);
+                    mainPathNodes.add(frameState);
+                }
+            }
+            while (!workList.isEmpty()) {
+                final Node n = workList.pop();
+                if (n instanceof PhiNode) {
+                    // don't follow inputs
+                } else if (n instanceof MergeNode) {
+                    // dont' follow inputs
+                } else {
+                    for (Node input : n.inputs()) {
+                        if (loop.whole().contains(input) &&
+                                        !mainPathNodes.contains(input)) {
+                            assert !workList.contains(input);
+                            workList.push(input);
+                            mainPathNodes.add(input);
+                        }
+                    }
+                }
+            }
+            if (cur == begin) {
+                break;
+            }
+            cur = cur.getDominator();
+        }
+
+        assert workList.isEmpty();
+        return mainPathNodes;
+    }
+
+    // Create a range check predicate for
+    //
+    // for (i = init; i < limit; i += stride) {
+    // a[scale*i+offset]
+    // }
+    private static void processGuard(LoopEx loop, GuardNode guard) {
+        final LogicNode condition = guard.getCondition();
+        if (!(condition instanceof IntegerBelowNode) || guard.isNegated()) {
+            return;
+        }
+
+        final IntegerBelowNode rangeCheck = (IntegerBelowNode) condition;
+        final ValueNode range = rangeCheck.getY();
+        if (!loop.isOutsideLoop(range) || ((IntegerStamp) range.stamp(NodeView.DEFAULT)).lowerBound() < 0) {
+            return;
+        }
+
+        final EconomicMap<Node, InductionVariable> inductionVariables = loop.getInductionVariables();
+
+        final ValueNode x = rangeCheck.getX();
+        if (!inductionVariables.containsKey(x)) {
+            return;
+        }
+
+        final StructuredGraph graph = guard.graph();
+
+        final InductionVariable iv = inductionVariables.get(x);
+
+        Long scale = null;
+
+        final InductionVariable counter = loop.counted().getCounter();
+        if (iv.isConstantScale(counter)) {
+            scale = iv.constantScale(counter);
+        }
+
+        ValueNode offset = null;
+
+        if (iv.offsetIsZero(counter)) {
+            offset = graph.unique(ConstantNode.forInt(0));
+        } else {
+            offset = iv.offsetNode(counter);
+        }
+
+        if (offset == null || scale == null || !loop.isOutsideLoop(offset)) {
+            return;
+        }
+
+        long scaleCon = scale;
+
+        final GuardNode upperGuard = getGuardNode(loop, guard, range, graph, scaleCon, offset, true);
+        final GuardNode lowerGuard = getGuardNode(loop, guard, range, graph, scaleCon, offset, false);
+
+        GuardingNode combinedGuard = null;
```

Comment:
Could we just do
```
guard.replaceAtUsagesAndDelete(MultiGuardNode.combine(lowerGuard, upperGuard));
```
and let `MultiGuardNode` combine if they want to during canonicalization.
In particular we need to create the `MultiGuardNode` regardless of whether the `guard` has usages.

Reply:
OK. A MultiGuardNode with MultiGuardNode inputs is not canonicalized right now AFAICT. Does that need to be changed?